pyrdp/player/gdi/cache.py

```diff
@@ -0,0 +1,202 @@
+#
+# This file is part of the PyRDP project.
+# Copyright (C) 2019 GoSecure Inc.
+# Licensed under the GPLv3 or later.
+#
+
+"""
+GDI Cache Management Layer.
+"""
+
+from PySide2.QtCore import QSize
+from PySide2.QtGui import QBrush, QImage, QBitmap
+
+from pyrdp.parser.rdp.orders.common import Glyph
+
+
+class BitmapCache:
+    """Bitmap cache."""
+
+    def __init__(self, persist=False):
+        self.caches = {}
+        self.persist = persist
+        if persist:
+            raise Exception('Persistent cache is not supported yet.')
+
+    def has(self, cid: int, idx: int) -> bool:
+        """
+        Check whether a cache contains an entry.
+
+        :param cid: The cache id to use.
+        :param idx: The cache entry index.
+
+        :returns: True if (cid:idx) is in the cache, false otherwise.
+        """
+        if cid not in self.caches:
+            return False
+        cache = self.caches[cid]
+        return idx in cache
+
+    def get(self, cid: int, idx: int) -> QImage:
+        """
+        Retrieve an entry from the cache.
+
+        :param cid: The cache id to use.
+        :param idx: The cache entry index.
+
+        :returns: The cache entry or None if it does not exist.
+        """
+        if cid not in self.caches:
+            return None
+        cache = self.caches[cid]
+        if idx not in cache:
+            return None
+        return cache[idx]
+
+    def add(self, cid: int, idx: int, entry: QImage) -> bool:
+        """
+        Add an entry to the cache.
+
+        :returns: True if the entry is a fresh entry, False if it replaced an existing one.
+        """
+        if cid not in self.caches:
+            self.caches[cid] = {}
+        cache = self.caches[cid]
+        cache[idx] = entry
+
+    def evict(self, cid: int, idx: int) -> bool:
```

Comment:
There is actually a maximum number of cache entries which get recycled. I used a dictionary in Python because it's a lot easier (and cleaner) than managing bounded arrays. Eviction is requested by the server, but once we drop a pane, all of this code should get garbage collected. A single connection is bound by how much cache entries it can have, so there is an upper limit to the memory consumption.

This should not be a problem.